File: Go/ds/CircularLinkedList.go
package main

import (
	"errors"
	"fmt"
	"math/rand"
)

// A linked list node
type node struct {
	data interface{}
	next *node
}

// A circular linked list
type circularLinkedList struct {
	head *node
	tail *node
	size int
}

// Inserts the given data to the head of the circular linked list
func (list *circularLinkedList) insert(value interface{}) {
	var temp = node{
		data: value,
	}

	if list.head == nil {
		list.head = &temp
		list.tail = &temp
		list.tail.next = list.head
	} else {
		list.tail.next = &temp
		list.tail = &temp
		list.tail.next = list.head
	}

	list.size++
}

// Prints the circular linked list
func (list *circularLinkedList) display() {
	var curr = list.head

	if list.head == nil {
		fmt.Println("Empty list")
	} else {
		fmt.Print("Nodes of Circular Linked List are: ")

		for curr.next != list.head {
			fmt.Print(curr.data)
			fmt.Print(" ")
			curr = curr.next
		}
		fmt.Print(curr.data)

		fmt.Println()
		fmt.Println(list.size)
	}
}

// Remove the node, if present, at the given index
func (list *circularLinkedList) remove(index int) (interface{}, error) {
	if index >= list.size {
		return nil, errors.New("Index not present")
	}

	list.size--
	if index == 0 {
		var temp = list.head
		list.head = temp.next
		list.tail.next = list.head

		return temp.data, nil
	}

	var iter = 0
	var curr = list.head

	for iter+1 < index {
		iter++
		fmt.Println(curr.data)
		curr = curr.next
	}
	var temp = curr
	curr.next = temp.next.next

	return temp.next.data, nil
}

// Returns the size of the given circular linked list
func (list *circularLinkedList) length() int {
	return list.size
}

// Returns a random element to insert in the Circular Linked List
func getElement() interface{} {
	var temp interface{} = rand.Intn(100)

	return temp
}

// Returns a random index to remove from the Circular Linked List
func getIndex(size int) int {
	var temp int = rand.Intn(100) % size

	return temp
}

func main() {
	var list circularLinkedList
	var choice int
	var flag bool

	flag = true

	for flag {
		fmt.Print("Menu\n1. Insert element\n2. Remove element\n3. Display elements\n4. Count of elements\n5. Exit\nEnter your choice: ")
		fmt.Scan(&choice)

		switch choice {
		case 1:
			element := getElement()
			list.insert(element)
			fmt.Println("Inserted: ", element)
		case 2:
			if list.size == 0 {
				fmt.Println("No element present in the list.")
			} else {
				index := getIndex(list.size)
				data, err := list.remove(index)
				if err == nil {
					fmt.Println("Removed: ", data)
				} else {
					fmt.Println("Error: ", err)
				}
			}
		case 3:
			list.display()
		case 4:
			count := list.length()
			fmt.Println("Length: ", count)
		case 5:
			flag = false
		default:
			fmt.Println("Please enter a valid choice.")
		}
	}
}

File: C/ds/CircularLinkedList.c
#include<stdio.h>
#include<conio.h>

void insertAtBeginning(int);
void insertAtEnd(int);
void insertAtAfter(int,int);
void deleteBeginning();
void deleteEnd();
void deleteSpecific(int);
void display();

struct Node
{
   int data;
   struct Node *next;
}*head = NULL;

void main()
{
   int choice1, choice2, value, location;
   clrscr();
   while(1)
   {
      printf("\n*********** MENU *************\n");
      printf("1. Insert\n2. Delete\n3. Display\n4. Exit\nEnter your choice: ");
      scanf("%d",&choice1);
      switch()
      {
         case 1: printf("Enter the value to be inserted: ");
         		 scanf("%d",&value);
                 while(1)
                 {
                 	printf("\nSelect from the following Inserting options\n");
                 	printf("1. At Beginning\n2. At End\n3. After a Node\n4. Cancel\nEnter your choice: ");
                    scanf("%d",&choice2);
                    switch(choice2)
                    {
                       case 1: 	insertAtBeginning(value);
                       		break;
                       case 2: 	insertAtEnd(value);
                       		break;
                       case 3: 	printf("Enter the location after which you want to insert: ");
                       		scanf("%d",&location);
                       		insertAfter(value,location);
                       		break;
                       case 4: 	goto EndSwitch;
                       default: printf("\nPlease select correct Inserting option!!!\n");
                    }
                 }
         case 2: while(1)
                 {
                 	printf("\nSelect from the following Deleting options\n");
                 	printf("1. At Beginning\n2. At End\n3. Specific Node\n4. Cancel\nEnter your choice: ");
                    scanf("%d",&choice2);
                    switch(choice2)
                    {
                       case 1: 	deleteBeginning();
                       		break;
                       case 2: 	deleteEnd();
                       		break;
                       case 3: 	printf("Enter the Node value to be deleted: ");
                       		scanf("%d",&location);
                       		deleteSpecic(location);
                       		break;
                       case 4:	goto EndSwitch;
                       default: printf("\nPlease select correct Deleting option!!!\n");
                    }
                 }
                 EndSwitch: break;
         case 3: display();
         	 break;
         case 4: exit(0);
         default: printf("\nPlease select correct option!!!");
      }
   }
}

void insertAtBeginning(int value)
{
    struct Node *newNode;
    newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode -> data = value;
    if(head == NULL)
    {
       head = newNode;
       newNode -> next = head;
    }
    else
    {
       struct Node *temp = head;
       while(temp -> next != head)
          temp = temp -> next;
       newNode -> next = head;
       head = newNode;
       temp -> next = head;
    }
    printf("\nInsertion success!!!");
}
void insertAtEnd(int value)
{
   struct Node *newNode;
   newNode = (struct Node*)malloc(sizeof(struct Node));
   newNode -> data = value;
   if(head == NULL)
   {
      head = newNode;
      newNode -> next = head;
   }
   else
   {
      struct Node *temp = head;
      while(temp -> next != head)
         temp = temp -> next;
      temp -> next = newNode;
      newNode -> next = head;
   }
   printf("\nInsertion success!!!");   
}
void insertAfter(int value, int location)
{
   struct Node *newNode;
   newNode = (struct Node*)malloc(sizeof(struct Node));
   newNode -> data = value;
   if(head == NULL)
   {      
      head = newNode;
      newNode -> next = head;
   }
   else
   {
      struct Node *temp = head;
      while(temp -> data != location)
      {
         if(temp -> next == head)
         {
            printf("Given node is not found in the list!!!");
            goto EndFunction;
         }
         else
         {
            temp = temp -> next;
         }
      }
      newNode -> next = temp -> next;
      temp -> next = newNode;
      printf("\nInsertion success!!!");
   }
   EndFunction:
}
void deleteBeginning()
{
   if(head == NULL)
      printf("List is Empty!!! Deletion not possible!!!");
   else
   {
      struct Node *temp = head;
      if(temp -> next == head)
      {
         head = NULL;
         free(temp);
      }
      else{
         head = head -> next;
         free(temp);
      }
      printf("\nDeletion success!!!");
   }
}
void deleteEnd()
{
   if(head == NULL)
      printf("List is Empty!!! Deletion not possible!!!");
   else
   {
      struct Node *temp1 = head, temp2;
      if(temp1 -> next == head)
      {
         head = NULL;
         free(temp1);
      }
      else{
         while(temp1 -> next != head){
            temp2 = temp1;
            temp1 = temp1 -> next;
         }
         temp2 -> next = head;
         free(temp1);
      }
      printf("\nDeletion success!!!");
   }
}
void deleteSpecific(int delValue)
{
   if(head == NULL)
      printf("List is Empty!!! Deletion not possible!!!");
   else
   {
      struct Node *temp1 = head, temp2;
      while(temp1 -> data != delValue)
      {
         if(temp1 -> next == head)
         {
            printf("\nGiven node is not found in the list!!!");
            goto FuctionEnd;
         }
         else
         {
            temp2 = temp1;
            temp1 = temp1 -> next;
         }
      }
      if(temp1 -> next == head){
         head = NULL;
         free(temp1);
      }
      else{
         if(temp1 == head)
         {
            temp2 = head;
            while(temp2 -> next != head)
               temp2 = temp2 -> next;
            head = head -> next;
            temp2 -> next = head;
            free(temp1);
         }
         else
         {
            if(temp1 -> next == head)
            {
               temp2 -> next = head;
            }
            else
            {
               temp2 -> next = temp1 -> next;
            }   
            free(temp1);
         }
      }
      printf("\nDeletion success!!!");
   }
   FuctionEnd:
}
void display()
{
   if(head == NULL)
      printf("\nList is Empty!!!");
   else
   {
      struct Node *temp = head;
      printf("\nList elements are: \n");
      while(temp -> next != head)
      {
         printf("%d ---> ",temp -> data);
      }
      printf("%d ---> %d", temp -> data, head -> data);
   }
}